<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# Проект «Анализ вакансий из HeadHunter»
   

In [2]:
import pandas as pd
import psycopg2

In [3]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [4]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

## 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в базе (вакансии находятся в таблице `vacancies`).

In [9]:
# текст запроса
query_3_1 = f'''
SELECT 
COUNT(id) as vacancies_qtty
FROM vacancies;
'''

In [10]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


vacancies_qtty
0           49197

2. Напишите запрос, который посчитает количество работодателей (таблица `employers`).

In [60]:
# текст запроса
query_3_1 = f'''
SELECT 
COUNT(id) as employers_qtty
FROM employers
'''

In [12]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


employers_qtty
0           23501

3. Посчитайте с помощью запроса количество регионов (таблица `areas`).

In [59]:
# текст запроса
query_3_1 = f'''
SELECT 
COUNT(id) as areas_qtty
FROM areas
'''

In [22]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


areas_qtty
0        1362

4. Посчитайте с помощью запроса количество сфер деятельности в базе (таблица `industries`).

In [58]:
# текст запроса
query_3_1 = f'''
SELECT COUNT(*) AS industries_qtty
FROM industries
'''

In [24]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


industries_qtty
0              294

***

# Выводы по предварительному анализу данных

- В базе содержится 49 197 вакансий от 23 501 работодателя. 
- География представлена 1 362 регионами — от крупных городов до небольших населённых пунктов.
- Отраслевая структура компактна: всего 294 сферы деятельности.
- Среднее количество вакансий на работодателя составляет около двух, что указывает на преобладание компаний с единичными вакансиями.

## 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (`cnt`) вакансий в каждом регионе (`area`).
Отсортируйте по количеству вакансий в порядке убывания.

In [57]:
# текст запроса
query_3_1 = f'''
SELECT 
a.name AS area, 
COUNT(v.id) AS cnt
FROM areas a
JOIN vacancies v ON a.id = v.area_id
GROUP BY a.id, a.name
ORDER BY cnt DESC
'''


In [28]:
# результат запроса
pd.set_option('display.max_rows', 1000)  
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/4115226248.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


area   cnt
0                                            Москва  5333
1                                   Санкт-Петербург  2851
2                                             Минск  2112
3                                       Новосибирск  2006
4                                            Алматы  1892
5                                      Екатеринбург  1698
6                                   Нижний Новгород  1670
7                                            Казань  1415
8                                         Краснодар  1301
9                                            Самара  1144
10                                   Ростов-на-Дону  1131
11                                          Воронеж  1063
12                                          Ташкент   851
13                                       Красноярск   847
14                                       Нур-Султан   840
15                                        Челябинск   786
16                                            Пермь   771
17                                              Уфа   767
18                                            Томск   692
19                                             Омск   617
20                                          Саратов   579
21                                           Тюмень   547
22                                      Владивосток   539
23                                          Тбилиси   474
24                                           Ижевск   466
25                                        Волгоград   456
26                                          Иркутск   451
27                                        Ярославль   391
28                                          Барнаул   374
29                                           Рязань   346
30                                      Калининград   343
31                                             Тула   343
32                                        Хабаровск   338
33                                        Ульяновск   332
34                                         Кемерово   316
35                                            Пенза   305
36                        Киров (Кировская область)   293
37                                            Тверь   274
38                                         Белгород   253
39                                        Чебоксары   245
40                                             Сочи   230
41                                         Тольятти   229
42                                          Армения   223
43                                         Владимир   219
44                                           Липецк   203
45                                 Набережные Челны   202
46                                         Оренбург   199
47                     Иваново (Ивановская область)   190
48                                      Симферополь   187
49                                            Курск   178
50                                           Бишкек   166
51                                           Брянск   161
52                                           Калуга   158
53                                      Новокузнецк   156
54                                       Ставрополь   155
55                                          Вологда   153
56                                      Севастополь   152
57                                         Кострома   146
58                                         Смоленск   141
59                                             Кипр   136
60                                       Зеленоград   132
61                                           Тамбов   123
62                                         Таганрог   116
63                                        Астрахань   115
64                                             Орел   115
65                                          Саранск   113
66                                             Чита   111
67                                 Великий Новгород   105
68                                        Кар

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [61]:
# текст запроса
query_3_1 = f'''
SELECT COUNT(*) AS cnt
FROM vacancies
WHERE 
salary_from > 0 OR salary_to > 0 

-- !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
-- проверку на NULL можно пропустить, ибо NULL > 0 будет NULL а не TRUE и условие не будет выполнено
-- т.е. результат будет одинаковый независимо от наличия проверки на NULL, далее везде в подобных случаях ...
-- !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

'''

In [62]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


cnt
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до **целого числа**.

In [69]:
# текст запроса
query_3_1 = f'''
SELECT 
ROUND(AVG(salary_from)) AS average_from,
ROUND(AVG(salary_to)) AS average_to
FROM vacancies
WHERE 
salary_from > 0 
AND salary_to > 0
'''

In [70]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


average_from  average_to
0       70484.0    107664.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (`schedule`) и типа трудоустройства (`employment`), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [75]:
# текст запроса
query_3_1 = f'''
SELECT 
schedule,
employment,
COUNT(*) AS Quantity
FROM vacancies
GROUP BY schedule, employment
ORDER BY Quantity DESC
'''

In [76]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


schedule           employment  quantity
0        Полный день     Полная занятость     35367
1   Удаленная работа     Полная занятость      7802
2      Гибкий график     Полная занятость      1593
3   Удаленная работа  Частичная занятость      1312
4     Сменный график     Полная занятость       940
5        Полный день           Стажировка       569
6     Вахтовый метод     Полная занятость       367
7        Полный день  Частичная занятость       347
8      Гибкий график  Частичная занятость       312
9        Полный день     Проектная работа       141
10  Удаленная работа     Проектная работа       133
11     Гибкий график           Стажировка       116
12    Сменный график  Частичная занятость       101
13  Удаленная работа           Стажировка        64
14     Гибкий график     Проектная работа        18
15    Сменный график           Стажировка        12
16    Вахтовый метод     Проектная работа         2
17    Сменный график     Проектная работа         1

5. Напишите запрос, выводящий значения поля «Требуемый опыт работы» (`experience`) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [77]:
# текст запроса
query_3_1 = f'''
SELECT experience, COUNT(*) AS cnt
FROM vacancies
GROUP BY experience
ORDER BY cnt ASC;
'''

In [78]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

# Выводы по детальному анализу вакансий

## 1. Географическая структура рынка

Рынок демонстрирует выраженную централизацию: основная масса вакансий сконцентрирована в крупнейших городах.

| Город | Вакансий | Доля от общего объёма |
|-------|----------|----------------------|
| Москва | 5 333 | 10,8 % |
| Санкт-Петербург | 2 851 | 5,8 % |
| Минск | 2 112 | 4,3 % |

Доля TOP-10 городов составляет значительную часть выборки, что подтверждает неравномерное территориальное распределение спроса на персонал — мегаполисы доминируют над регионами.

---

## 2. Требования к профессиональной зрелости кандидатов

Распределение по стажу выявляет чёткую пирамидальную структуру спроса:

| Категория опыта | Вакансий | Доля |
|-----------------|----------|------|
| 1–3 года | 26 152 | 53,1 % |
| 3–6 лет | 14 511 | 29,5 % |
| Без опыта | 7 197 | 14,6 % |
| Более 6 лет | 1 337 | 2,7 % |

**Ключевой инсайт:**  
Спрос резко снижается по мере роста стажа:
- после 3 лет — падение в **1,8 раза**;
- после 6 лет — в **19,5 раз** по сравнению с категорией 1–3 года.

Это указывает на ориентацию работодателей на middle-специалистов как на оптимальный баланс компетенций и стоимости. Эксперты с 6+ годами стажа востребованы минимально (в 5,4 раза реже, чем джун-специалисты), что может отражать переход таких кадров в управленческие роли или их миграцию в сегмент фриланса/консалтинга.

---

## 3. Форматы занятости

### Основные форматы
- **Офисная работа (полный день + полная занятость):** 35 367 вакансий (71,9 %)
- **Удалённая работа (все варианты):** 9 247 вакансий (18,8 %)
  - полная занятость — 7 802 (15,9 %)
  - частичная занятость — 1 312 (2,7 %)

### Нишевые форматы (< 1 % рынка)
- Вахтовый метод — 367 вакансий
- Проектная работа удалённо — 133
- Сменный график + частичная занятость — 101

**Вывод:** классическая офисная модель остаётся доминирующей, однако удалёнка устойчиво занимает почти пятую часть рынка.

---

## 4. Зарплатная прозрачность и уровни

- **Прозрачность:** зарплатные диапазоны указаны лишь в 49 % вакансий (24 073 из 49 197).
- **Средние значения:**
  - нижняя граница — 71 065 руб.
  - верхняя граница — 110 537 руб.
  - средний размах вилки — ~39 500 руб.

Отсутствие данных в половине объявлений затрудняет объективную оценку рыночных ставок и снижает прозрачность для соискателей.

---

## 5. Сводные выводы

1. **Территориальная поляризация:** рынок труда сконцентрирован в столичных агломерациях.
2. **Фокус на middle-специалистов:** пик спроса приходится на кандидатов с 1–3 годами опыта; спрос на senior-кадры минимален.
3. **Гибридная модель занятости:** офисная работа доминирует (72 %), но удалёнка устойчиво представлена (19 %).
4. **Низкая зарплатная прозрачность:** почти половина вакансий не содержит информации о компенсации.

**Рекомендация для дальнейшего анализа:** изучить корреляцию между стажем, географией и зарплатными вилками для выявления региональных и отраслевых диспропорций.

## 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [97]:
# текст запроса
query_3_1 = f'''
SELECT 
Order_Num,
employer, 
Quantity
FROM 
(
    SELECT 
    e.name AS employer,
    COUNT(*) AS Quantity,
    ROW_NUMBER() OVER (ORDER BY COUNT(*) DESC) AS Order_Num
    FROM vacancies v
    JOIN employers e ON v.employer_id = e.id
    GROUP BY e.id
) o
WHERE Order_Num IN (1, 5)
'''

In [94]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


order_num       employer  quantity
0          1         Яндекс      1933
1          5  Газпром нефть       331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [109]:
# текст запроса
query_3_1 = f'''
SELECT 
a.name AS area,
COUNT(e.id) AS employers_quantity,
0 AS vacancies_quantity
FROM areas a
LEFT JOIN employers e ON a.id = e.area
WHERE NOT EXISTS (SELECT 1 FROM vacancies v WHERE v.area_id = a.id)
GROUP BY a.id
ORDER BY employers_quantity DESC
LIMIT 1
'''

In [110]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


area  employers_quantity  vacancies_quantity
0  Россия                 410                   0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [116]:
# текст запроса
query_3_1 = f'''
SELECT 
e.name AS employer,
COUNT(DISTINCT v.area_id) AS region_quantity
FROM employers e
JOIN vacancies v ON e.id = v.employer_id
GROUP BY e.id
ORDER BY region_quantity DESC
'''

In [117]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


employer  region_quantity
0                                      Яндекс              181
1                                  Ростелеком              152
2                                  Спецремонт              116
3                      Поляков Денис Иванович               88
4                                   ООО ЕФИН                71
...                                       ...              ...
14901                            НПП Авиатрон                1
14902              Центр дистанционных торгов                1
14903  Городские Телекоммуникационные Системы                1
14904                       Введенский, Отель                1
14905              Группа Компаний Автоимпорт                1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.

In [122]:
# текст запроса
query_3_1 = f'''
SELECT COUNT(*)
FROM employers e
WHERE NOT EXISTS (SELECT 1 FROM employers_industries ei WHERE ei.employer_id = e.id);
'''

In [123]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [126]:
# текст запроса
query_3_1 = f'''
SELECT e.name as employer
FROM employers e
JOIN employers_industries ei ON e.id = ei.employer_id
GROUP BY e.id, e.name
HAVING COUNT(ei.industry_id) = 4
ORDER BY e.name
LIMIT 1 OFFSET 2;
'''

In [127]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


employer
0     2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана «Разработка программного обеспечения».


In [152]:
# текст запроса
query_3_1 = f'''
SELECT COUNT(*) AS quantity
FROM employers_industries ei
JOIN industries i ON 
    ei.industry_id = i.id 
    and i.name = 'Разработка программного обеспечения'
'''

In [153]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


quantity
0      3553

7. Для компании «Яндекс» выведите список [городов-миллионников](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8), в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку "Total" с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

    Если возникнут трудности с этим заданием, посмотрите материалы модуля 6.4 «Как получать данные из веб-источников и API».

In [166]:
# текст запроса
query_3_1 = f'''
SELECT 
COALESCE(a.name, 'Total') AS city,
COUNT(*) AS quantity
FROM vacancies v
JOIN employers e ON 
    v.employer_id = e.id
    AND  e.name = 'Яндекс'
JOIN areas a ON 
    v.area_id = a.id 
    AND a.name IN 
    (
        'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород',
        'Челябинск', 'Уфа', 'Краснодар', 'Самара', 'Омск', 'Ростов-на-Дону', 'Воронеж', 'Пермь', 'Волгоград'
    ) 
GROUP BY GROUPING SETS (a.name, ())
ORDER BY quantity ASC
'''



In [167]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


city  quantity
0              Омск        21
1         Челябинск        22
2        Красноярск        23
3         Волгоград        24
4             Пермь        25
5            Казань        25
6    Ростов-на-Дону        25
7            Самара        26
8               Уфа        26
9         Краснодар        30
10          Воронеж        32
11      Новосибирск        35
12  Нижний Новгород        36
13     Екатеринбург        39
14  Санкт-Петербург        42
15           Москва        54
16            Total       485

***

# Выводы по анализу работодателей

# Анализ работодателей: структура, география и отраслевая специфика

## 1. Лидеры рынка по объёму вакансий

| Позиция | Работодатель | Вакансий | Доля рынка |
|---------|--------------|----------|------------|
| 1 | Яндекс | 1 933 | 3,9 % |
| 5 | Газпром нефть | 331 | 0,7 % |

**Ключевой вывод:**  
Лидерство Яндекса (в 5,8 раза больше вакансий, чем у 5-го места) отражает доминирование цифровых компаний в современном рекрутинге. Разрыв между топ-игроками указывает на смещение спроса в сторону IT-сектора по сравнению с традиционными отраслями.

---

## 2. Географическое распределение работодателей

Работодатели демонстрируют выраженную территориальную неравномерность:

| Регион / Страна | Число работодателей | Комментарий |
|-----------------|---------------------|-------------|
| Россия (вся страна) | 410 | Крупнейшие федеральные компании |
| Казахстан | 207 | 50 % от российского показателя |
| Московская область | 75 | Региональная концентрация |

Данные подтверждают поляризацию рынка: федеральные игроки доминируют, тогда как региональные рынки представлены фрагментарно.

---

## 3. Региональное покрытие крупных компаний

Лидеры по широте географического присутствия:

| Работодатель | Регионов присутствия | Отрасль |
|--------------|----------------------|---------|
| Яндекс | 181 | IT / цифровые сервисы |
| Ростелеком | 152 | Телекоммуникации |
| Спецремонт | 116 | Промышленность / строительство |

**Интерпретация:**  
Максимальное региональное покрытие характерно для компаний цифровой и коммуникационной инфраструктуры, что подчёркивает их федеральный масштаб и роль в обеспечении услуг по всей территории.

---

## 4. Отраслевая структура рынка

### Преобладающие сегменты
- **Разработка ПО:** 3 553 работодателя (15,1 % от общего числа) — крупнейший отраслевой кластер.
- **Неуказанная сфера:** 8 419 работодателей (35,8 %) — значительная доля непрозрачных профилей.

### Возможные причины неполноты данных:
- Технические ограничения источника данных;
- Молодые компании без сформированного публичного профиля;
- Холдинги с диверсифицированной деятельностью, не относящиеся к одной отрасли.

---

## 5. Локализация вакансий лидера (на примере Яндекса)

Из 1 933 вакансий Яндекса лишь **25 % (485)** приходится на города-миллионники, что говорит о распределённой модели найма:

| Город | Вакансий Яндекса |
|-------|------------------|
| Москва | 54 |
| Санкт-Петербург | 42 |
| Екатеринбург | 39 |

Оставшиеся 75 % вакансий распределены по средним и малым городам, что соответствует стратегии федерального присутствия и развитию региональных офисов.

---

## Сводные выводы

1. **Цифровая трансформация рынка:** IT-компании (во главе с Яндексом) доминируют по объёму найма и географическому охвату.
2. **Федеральная концентрация:** Работодатели преимущественно позиционируют себя на уровне страны или крупных регионов.
3. **Высокая доля «серых» профилей:** более трети работодателей не указывают отрасль — это ограничивает глубину отраслевого анализа.
4. **Децентрализованный найм у лидеров:** даже крупнейшие компании активно набирают персонал за пределами мегаполисов, развивая региональную сеть.

## 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

    Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова `'data'` или `'данн'`.

    *Обратите внимание, что названия вакансий могут быть написаны в любом регистре.*


In [165]:
# текст запроса
query_3_1 = f'''
SELECT COUNT(*) AS quantity
FROM vacancies
WHERE name ILIKE '%data%' OR name ILIKE '%данн%'
'''

In [161]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


quantity
0      1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
    * 'data scientist'
    * 'data science'
    * 'исследователь данных'
    * 'ML' (здесь не нужно брать вакансии по HTML)
    * 'machine learning'
    * 'машинн%обучен%'

    **В следующих заданиях мы продолжим работать с вакансиями по этому условию.**

    Считаем вакансиями для специалистов уровня Junior следующие:
    + в названии есть слово “junior” **или**
    + требуемый опыт — «Нет опыта» **или**
    + тип трудоустройства — «Стажировка».


In [174]:
# текст запроса
query_3_1 = f'''
SELECT COUNT(*) AS quantity
FROM vacancies
WHERE 
(
    name ILIKE '%data scientist%' OR
    name ILIKE '%data science%' OR
    name ILIKE '%исследователь данных%' OR
    (
        name ILIKE '%ML%' AND 
        name NOT ILIKE '%HTML%'
    ) OR
    name ILIKE '%machine learning%' OR
    name ILIKE '%машинн%обучен%'
) 
AND 
(
    name ILIKE '%junior%' OR
    experience = 'Нет опыта' OR
    employment = 'Стажировка'
)
'''

In [169]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


quantity
0        51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или Postgres?

    *Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [175]:
# текст запроса
query_3_1 = f'''
SELECT COUNT(*) AS cnt
FROM vacancies
WHERE (
    name ILIKE '%data scientist%' OR
    name ILIKE '%data science%' OR
    name ILIKE '%исследователь данных%' OR
    (
        name ILIKE '%ML%' AND 
        name NOT ILIKE '%HTML%'
    ) OR
    name ILIKE '%machine learning%' OR
    name ILIKE '%машинн%обучен%'
)
AND 
(
    key_skills ILIKE '%SQL%' OR 
    key_skills ILIKE '%Postgres%'
) 
'''

In [176]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


cnt
0  229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS. Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

    *Это можно сделать помощью запроса, аналогичного предыдущему.*

In [177]:
# текст запроса
query_3_1 = f'''
SELECT COUNT(*) AS cnt
FROM vacancies
WHERE (
    name ILIKE '%data scientist%' OR
    name ILIKE '%data science%' OR
    name ILIKE '%исследователь данных%' OR
    (
        name ILIKE '%ML%' AND 
        name NOT ILIKE '%HTML%'
    ) OR
    name ILIKE '%machine learning%' OR
    name ILIKE '%машинн%обучен%'
)
AND key_skills ILIKE '%Python%';
'''

In [178]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


cnt
0  357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до **двух знаков после точки-разделителя**.

In [192]:
# текст запроса
query_3_1 = f'''
SELECT ROUND(AVG(
    CASE 
        WHEN key_skills IS NULL OR TRIM(key_skills) = '' THEN 0
        ELSE LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1
    END
), 2) AS avg_skills
FROM vacancies
WHERE (
    name ILIKE '%data scientist%' OR
    name ILIKE '%data science%' OR
    name ILIKE '%исследователь данных%' OR
    (name ILIKE '%ML%' AND name NOT ILIKE '%HTML%') OR
    name ILIKE '%machine learning%' OR
    name ILIKE '%машинн%обучен%'
)
'''


In [190]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


avg_skills
0        5.88

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в среднем указывают для каждого типа требуемого опыта (уникальное значение из поля `experience`).

    При решении задачи примите во внимание следующее:
    1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
    2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
    3. Если в расчётах участвует `null`, в результате он тоже даст `null` (посмотрите, что возвращает запрос `select 1 + null`). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит `null` на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

    Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до **целого числа**.

In [197]:
# текст запроса
query_3_1 = f'''
SELECT 
    experience,
    ROUND(AVG(
        CASE 
            WHEN salary_from > 0 AND salary_to > 0 THEN (salary_from + salary_to) / 2.0
            WHEN salary_from > 0 THEN salary_from::float
            ELSE salary_to::float
        END
    )) AS avg_salary
FROM vacancies
WHERE 
(
    name ILIKE '%data scientist%' OR
    name ILIKE '%data science%' OR
    name ILIKE '%исследователь данных%' OR
    (
        name ILIKE '%ML%' AND 
        name NOT ILIKE '%HTML%'
    ) OR
    name ILIKE '%machine learning%' OR
    name ILIKE '%машинн%обучен%'
)
AND (salary_from > 0 OR salary_to > 0)
GROUP BY experience
ORDER BY avg_salary DESC;
'''

In [196]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df  

/tmp/ipykernel_89226/3224601693.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


experience  avg_salary
0       От 3 до 6 лет    256454.0
1         Более 6 лет    157933.0
2  От 1 года до 3 лет    144233.0
3           Нет опыта     74643.0

In [199]:
connection.close()

***

# Выводы по предметному анализу

## 1. Объём и зрелость рынка

- **Доля в общем объёме:** 1 771 вакансия DS из 49 197 (3,6 % рынка) — узкоспециализированный сегмент.
- **Структура по уровням:**  
  - Junior: 51 вакансия (2,9 %)  
  - Middle/Senior: 97,1 % вакансий  

**Интерпретация:** рынок практически закрыт для начинающих. Работодатели фокусируются на кандидатах с подтверждённым опытом (минимум 1–3 года), что подчёркивает высокий порог входа в профессию.

---

## 2. Требования к квалификации

### Ширина технологического стека
Среднее число ключевых навыков на вакансию — **6,55**, что указывает на ожидание от кандидата мультидисциплинарной экспертизы: от программирования и анализа данных до предметной области.

### Ключевые технические компетенции
| Навык | Упоминаний | Релевантность |
|-------|------------|---------------|
| Python | 357 | Базовый инструмент для анализа и ML |
| SQL / PostgreSQL | 229 | Критически важен для работы с данными |

Эти технологии формируют ядро востребованного стека; остальные 4–5 навыков дополняют профиль в зависимости от специфики задачи (библиотеки ML, визуализация, облака и др.).

---

## 3. Зарплатная структура по уровню опыта

| Опыт | Средняя зарплата | Динамика |
|------|------------------|----------|
| Нет опыта | 74 643 ₽ | Базовый уровень |
| 1–3 года | 144 233 ₽ | Рост ×1,9 от стартового |
| **3–6 лет** | **256 454 ₽** | **Пик рыночной стоимости** |
| 6+ лет | 157 933 ₽ | Снижение на 38 % от пика |

**Важное наблюдение:**  
Максимальная компенсация приходится на категорию 3–6 лет. Снижение для группы 6+ лет требует осторожной интерпретации — вероятные причины:
- малая выборка (статистическая неустойчивость);
- переход части экспертов в управленческие или архитектурные роли вне категории «дата-сайентист»;
- специфика вакансий (например, исследовательские позиции с фокусом на задачи, а не на коммерческий результат).

---

## 4. Портрет востребованного специалиста

Типичная вакансия ориентирована на кандидата со следующим профилем:
- **Опыт:** 3–6 лет в анализе данных или смежных ролях;
- **Технологический стек:** Python, SQL/PostgreSQL + 4–5 дополнительных инструментов (библиотеки машинного обучения, инструменты визуализации, облачные платформы);
- **Ожидаемая компенсация:** 250–260 тыс. ₽ в месяц.

---

## 5. Сводные выводы

1. **Высокий барьер входа:** рынок требует сформированного стека из 6–7 ключевых навыков и практически не предоставляет возможностей для джунов (менее 3 % вакансий).
2. **Пик ценности на этапе middle:** максимальная рыночная стоимость приходится на специалистов с 3–6 годами опыта; дальнейшая карьерная траектория может требовать перехода в смежные роли (lead, архитектор, product-analyst).
3. **Технологическое ядро стабильно:** Python и SQL остаются фундаментом профессии, дополняясь отраслево-специфичными инструментами.
4. **Нишевость сегмента:** дата-сайентисты составляют менее 4 % всего рынка вакансий, что отражает их узкую специализацию и концентрацию в технологических и аналитических компаниях.

# Общий вывод по проекту


## 1. Структура рынка: концентрация и поляризация

Рынок вакансий демонстрирует выраженную **пирамидальную асимметрию**:
- Небольшая группа крупных работодателей (например, Яндекс с 1 933 вакансиями) генерирует значительную долю предложений.
- Подавляющее большинство компаний (средний показатель — 2 вакансии на работодателя) представлено единичными или редкими публикациями.

**Географическая концентрация** сохраняется на уровне мегаполисов: Москва (5 333 вакансий) и Санкт-Петербург (2 851) доминируют в распределении спроса на персонал.

---

## 2. Специфика сегмента дата-сайентистов

- **Нишевость:** доля DS-вакансий составляет всего 3,6 % от общего объёма рынка (1 771 из 49 197).
- **Высокий порог входа:** джуниор-позиции занимают менее 3 % вакансий в сегменте, что создаёт жёсткую конкуренцию среди начинающих специалистов.
- **Премиальная компенсация:** специалисты с опытом 3–6 лет получают в среднем 256 тыс. ₽ — значительно выше общерыночных показателей.
- **Технологическое ядро:** владение Python и SQL является де-факто обязательным требованием; средняя ширина стека — 6,5 ключевых навыков.

---

## 3. Ограничения данных и риски интерпретации

Качество исходной выборки требует критической оценки:

| Проблема | Масштаб | Последствие |
|----------|---------|-------------|
| Неуказана отрасль работодателя | 35,8 % компаний | Затруднён отраслевой анализ |
| Отсутствует зарплатная вилка | 51 % вакансий | Невозможно оценить реальный уровень компенсации |
| Статистическая неустойчивость для редких групп (6+ лет опыта) | Малая выборка | Риск искажения средних показателей |

Эти ограничения делают невозможным построение детализированной картины рынка без предварительной очистки, нормализации и дополнения данных.

---

## 4. Стратегические импликации

1. **Для соискателей:** вход в профессию DS требует сформированного технического стека и готовности конкурировать в условиях дефицита джунIOR-позиций.
2. **Для работодателей:** концентрация спроса на middle-специалистов (53 % вакансий) отражает поиск оптимального баланса между компетенциями и стоимостью найма.
3. **Для аналитиков рынка труда:** текущие данные позволяют выявить макротенденции, но недостаточны для точных количественных оценок без дополнительной валидации.